### PR TITLE
chore: Bump revm to newest

### PR DESCRIPTION
**Pull request by rakita**
Integrate revm to newest branch.

### CODE REVIEW

File: crates/payload/basic/src/lib.rs [hunk 1]
```diff
@@ -861,24 +864,23 @@ pub fn pre_block_beacon_root_contract_call<DB: Database + DatabaseCommit, Attrib
     db: &mut DB,
     chain_spec: &ChainSpec,
     block_number: u64,
-    initialized_cfg: &CfgEnv,
+    initialized_cfg: &CfgEnvWithHandlerCfg,
     initialized_block_env: &BlockEnv,
     attributes: &Attributes,
 ) -> Result<(), PayloadBuilderError>
 where
     DB::Error: std::fmt::Display,
```
Comment (emhane) [inline crates/payload/basic/src/lib.rs:872]: preferable if the trait bound is `std::error::Error`, no?

File: crates/node-api/src/evm/mod.rs [hunk 2]
```diff
@@ -1,5 +1,5 @@
 use reth_primitives::{revm::env::fill_block_env, Address, ChainSpec, Header, Transaction, U256};
-use revm_primitives::{BlockEnv, CfgEnv, SpecId, TxEnv};
+use revm_primitives::{BlockEnv, CfgEnvWithHandlerCfg, SpecId, TxEnv};
 
 /// This represents the set of methods used to configure the EVM before execution.
 pub trait EvmEnvConfig: Send + Sync + Unpin + Clone {
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/node-api/src/evm/mod.rs:5]: cc @Rjected this looks good

File: crates/payload/basic/src/lib.rs (hunk 1 above)
Comment (rakita) [reply crates/payload/basic/src/lib.rs:872]: I think this is okay, we have our own types. with standard Error we would just lose information

File: crates/node-api/src/evm/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Rjected) [reply crates/node-api/src/evm/mod.rs:5]: Thx for the guideline doc, definitely useful for naming conventions. I'll submit a PR for some trait renames

File: crates/primitives/src/log.rs [hunk 3]
```diff
@@ -20,6 +21,12 @@ pub struct Log {
     pub data: Bytes,
 }
 
+impl From<AlloyLog> for Log {
```
Comment (gakonst) [inline crates/primitives/src/log.rs:24]: Do we need this? Should we re-use types, or not possible? @mattsse @DaniPopes

File: crates/revm/src/processor.rs [hunk 4]
```diff
@@ -57,9 +56,7 @@ pub struct EVMProcessor<'a, EvmConfig> {
     /// The configured chain-spec
     pub(crate) chain_spec: Arc<ChainSpec>,
     /// revm instance that contains database and env environment.
-    pub(crate) evm: EVM<StateDBBox<'a, ProviderError>>,
-    /// Hook and inspector stack that we want to invoke on that hook.
-    stack: InspectorStack,
+    pub(crate) evm: Option<Evm<'a, InspectorStack, StateDBBox<'a, ProviderError>>>,
```
Comment (gakonst) [inline crates/revm/src/processor.rs:59]: Why is the EVM optional now?

File: crates/revm/src/processor.rs [hunk 5]
```diff
@@ -248,30 +260,46 @@ where
...
 
         let hash = transaction.hash();
-        let out = if self.stack.should_inspect(&self.evm.env, hash) {
+        let should_inspect = self
+            .evm
+            .as_ref()
+            .unwrap()
+            .context
+            .external
+            .should_inspect(self.evm.as_ref().unwrap().context.evm.env.as_ref(), hash);
```
Comment (gakonst) [inline crates/revm/src/processor.rs:284]: I don't love how there's now `self.evm.as_ref/mut().unwrap()` in multiple places now, can we hide this?

**Review by DaniPopes — CHANGES_REQUESTED** [review]

**Review by Rjected — CHANGES_REQUESTED** [review]
Main comments just have to do with the `Evm` being in an `Option` and reusing the `SpecId` if we already have it. The rest looks good to me!

EDIT: ah lol the Evm option was already brought up, too slow with my reviews 😅

File: crates/payload/basic/src/lib.rs [hunk 6]
```diff
@@ -852,7 +852,7 @@ pub fn commit_withdrawals<DB: Database<Error = ProviderError>>(
 
 /// Apply the [EIP-4788](https://eips.ethereum.org/EIPS/eip-4788) pre block contract call.
 ///
-/// This constructs a new [Evm](revm::Evm) with the given DB, and environment
+/// This constructs a new [Evm] with the given DB, and environment
```
Comment (emhane) [inline crates/payload/basic/src/lib.rs:855]:
```suggestion
/// This constructs a new [`Evm`](revm::Evm)  with the given DB, and environment
```

File: crates/rpc/rpc/src/eth/api/pending_block.rs [hunk 7]
```diff
@@ -267,7 +267,7 @@ impl PendingBlockEnv {
 
 /// Apply the [EIP-4788](https://eips.ethereum.org/EIPS/eip-4788) pre block contract call.
 ///
-/// This constructs a new [EVM](revm::Evm) with the given DB, and environment [CfgEnvWithHandlerCfg]
+/// This constructs a new [Evm](revm::Evm) with the given DB, and environment [CfgEnvWithHandlerCfg]
```
Comment (emhane) [inline crates/rpc/rpc/src/eth/api/pending_block.rs:270]:
```suggestion
/// This constructs a new [`Evm`](revm::Evm) with the given DB, and environment [`CfgEnvWithHandlerCfg`]
```

**Review by emhane — DISMISSED** [review]
readability improves with `code syntax` on linked types, traits, methods, etc.

File: crates/node-api/src/evm/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (emhane) [reply crates/node-api/src/evm/mod.rs:5]: nice, I will open an issue for renaming constants also

**Comment by gakonst** [comment]
Have we run @rkrasiuk debug scripts on tip for reorgs & merkle root?

File: crates/node-api/src/evm/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rakita) [reply crates/node-api/src/evm/mod.rs:5]: team effort: https://github.com/paradigmxyz/reth/issues/6420 lol

File: crates/primitives/src/log.rs (hunk 3 above)
Comment (rakita) [reply crates/primitives/src/log.rs:24]: Good guess @DaniPopes. Tried to switch it, but it didn't work, left it as it was.

File: crates/revm/src/processor.rs (hunk 4 above)
Comment (rakita) [reply crates/revm/src/processor.rs:59]: It was made so we can mutate `Evm` in `EvmBuilder` as I needed to change a generic of `external Context` to `InspectorStack`. But on the second iteration, I just embedded `InspectorStack` inside Evm always.

I should just add a few functions to `Handler` and there will be no need to mutate `Evm` and no need for `Option` 

cc @Rjected

[the remark this answers is not in the document]
Comment (rakita) [reply crates/revm/src/processor.rs:59]: cc'ed you on thread

File: crates/payload/basic/src/lib.rs [hunk 8]
```diff
@@ -630,10 +630,12 @@ impl Drop for Cancelled {
 /// Static config for how to build a payload.
 #[derive(Clone, Debug)]
 pub struct PayloadConfig<Attributes> {
+    /// Specification id for the block.
+    pub spec_id: SpecId,
     /// Pre-configured block environment.
     pub initialized_block_env: BlockEnv,
     /// Configuration for the environment.
-    pub initialized_cfg: CfgEnv,
+    pub initialized_cfg: CfgEnvWithHandlerCfg,
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/payload/basic/src/lib.rs:638]: It does, should remove `spec_id` as first iteration was with that field and with `CfgEnvWithHandlerCfg` it is not necessary.

**Comment by rakita** [comment]
> Have we run @rkrasiuk debug scripts on tip for reorgs & merkle root?

Nope as this does not touch merkle root

**Comment by rakita** [comment]
> > Have we run @rkrasiuk debug scripts on tip for reorgs & merkle root?
> 
> Nope as this does not touch merkle root. I did synced the Holesky

**Comment by rkrasiuk** [comment]
@rakita merkle scripts do not only verify the correctness of state root calculation, they also verify the correctness of the state. more specifically, i'd recommend running `debug execution` script since it also checks whether state reverts were written correctly

**Comment by rakita** [comment]
> @rakita merkle scripts do not only verify the correctness of state root calculation, they also verify the correctness of the state. more specifically, i'd recommend running `debug execution` script since it also checks whether state reverts were written correctly

revm State was not touched and had minimal changes. So revert creation and state change is the same and i didn't felt the need to run it as the fixes that we have were already merged in reth_freeze.

The best test for this change is syncing reth, and this, I already did with Holesky.

File: testing/ef-tests/tests/tests.rs [hunk 9]
```diff
@@ -12,6 +12,7 @@ macro_rules! general_state_test {
                 .stack_size(
```
[the remark this answers is not in the document]
Comment (rakita) [reply testing/ef-tests/tests/tests.rs:12]: removed here https://github.com/paradigmxyz/reth/commit/8ef88105ae975af8b8d186833da58e13244bd599

File: crates/payload/basic/src/lib.rs (hunk 6 above)
Comment (rakita) [reply crates/payload/basic/src/lib.rs:855]: No need for this

File: crates/rpc/rpc/src/eth/api/pending_block.rs (hunk 7 above)
Comment (rakita) [reply crates/rpc/rpc/src/eth/api/pending_block.rs:270]: No need for this

File: crates/node-api/src/engine/traits.rs [hunk 10]
```diff
@@ -131,7 +130,7 @@ pub trait PayloadBuilderAttributes: Send + Sync + std::fmt::Debug {
             blob_excess_gas_and_price,
         };
 
-        (cfg, block_env)
+        (spec_id, CfgEnvWithHandlerCfg::new(cfg, spec_id), block_env)
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/node-api/src/engine/traits.rs:133]: ref to https://github.com/paradigmxyz/reth/pull/6357#discussion_r1478849482

File: crates/payload/basic/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (rakita) [reply crates/payload/basic/src/lib.rs]: done

File: crates/revm/src/processor.rs (hunk 5 above)
Comment (rakita) [reply crates/revm/src/processor.rs]: removed with removal of Option

File: crates/revm/src/processor.rs (hunk 4 above)
Comment (rakita) [reply crates/revm/src/processor.rs]: done, the `Option` is removed. Added few functions directly to `revm::Handler`

**Review by DaniPopes — APPROVED** [review]
Code LGTM

File: crates/rpc/rpc/src/eth/api/pending_block.rs [hunk 11]
```diff
@@ -130,11 +131,13 @@ impl PendingBlockEnv {
             }
 
             // Configure the environment for the block.
-            let env =
-                Env { cfg: cfg.clone(), block: block_env.clone(), tx: tx_env_with_recovered(&tx) };
+            let env = Box::new(Env {
+                cfg: cfg.cfg_env.clone(),
+                block: block_env.clone(),
+                tx: tx_env_with_recovered(&tx),
+            });
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/pending_block.rs:138]: there should be a `fn boxed` on env

File: crates/rpc/rpc/src/eth/api/pending_block.rs [hunk 12]
```diff
@@ -130,11 +131,13 @@ impl PendingBlockEnv {
...
 
-            let mut evm = revm::EVM::with_env(env);
-            evm.database(&mut db);
+            let mut evm = revm::Evm::builder().modify_env(|e| *e = env).with_db(&mut db).build();
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/pending_block.rs:140]: there should be a function that accepts the env: `fn with_env`

File: crates/revm/src/processor.rs [hunk 13]
```diff
@@ -248,30 +257,40 @@ where
     ) -> Result<ResultAndState, BlockExecutionError> {
         // Fill revm structure.
         #[cfg(not(feature = "optimism"))]
-        fill_tx_env(&mut self.evm.env.tx, transaction, sender);
+        fill_tx_env(&mut self.evm.context.evm.env.tx, transaction, sender);
```
Comment (mattsse) [inline crates/revm/src/processor.rs:260]: there should be getters for this on evm imo

File: crates/rpc/rpc/src/eth/revm_utils.rs [hunk 14]
```diff
@@ -116,33 +116,45 @@ impl FillableTransaction for TransactionSigned {
...
-    let mut evm = revm::EVM::with_env(env);
-    evm.database(db);
-    let res = evm.inspect(inspector)?;
-    Ok((res, evm.env))
+    let mut evm = revm::Evm::builder()
+        .with_db(db)
+        .with_external_context(inspector)
+        .with_env_with_handler_cfg(env)
+        .append_handler_register(inspector_handle_register)
+        .build();
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/revm_utils.rs:154]: there's also `append_handle_register` somewhere else, not clear if this entirely different or if naming is not consistent

**Review by mattsse — CHANGES_REQUESTED** [review]
This is less invasive than expected, I think overall this lgtm, though I'm having some reservations wrt naming.

I'm being particularly pedantic here: 

It's a bit challenging to get familiar with the names, takes some time to get used to this for example:

there's `EnvWithHandlerCfg` and `CfgEnvWithHandlerCfg`

and `HandlerCfg` is just the wrapped SpecId, unclear what this has to do with a handler config, besides a feature gated optimism field. 
Can we get rid of optimism cfgs eventually by any chance?

there are no docs that cover what the nomenclature is here, 
I know they're most likely in the revm mdbook, but this is not very useful, should be directly in the code imo.

the evm type would benefit from some additional helper functions as well

although, I like that we don't have any generics on the handler type, this should make it easier to move this into a config trait

**Review by Rjected — APPROVED** [review]
LGTM!

**Comment by Rjected** [comment]
> there's `EnvWithHandlerCfg` and `CfgEnvWithHandlerCfg`
> 
> and `HandlerCfg` is just the wrapped SpecId, unclear what this has to do with a handler config, besides a feature gated optimism field.

Yeah, I found this pretty confusing too, worth considering renames

**Comment by Rjected** [comment]
oops meant to click "comment" not "close with comment"

File: crates/rpc/rpc/src/eth/revm_utils.rs (hunk 14 above)
Comment (rakita) [reply crates/rpc/rpc/src/eth/revm_utils.rs:154]: misspelled

File: crates/revm/src/processor.rs (hunk 13 above)
Comment (rakita) [reply crates/revm/src/processor.rs:260]: Added cfg,tx,block handlers and their `_mut` counterparts

File: crates/rpc/rpc/src/eth/api/pending_block.rs (hunk 12 above)
Comment (rakita) [reply crates/rpc/rpc/src/eth/api/pending_block.rs:140]: added `with_env`, `with_block`, `with_tx`

File: crates/rpc/rpc/src/eth/api/pending_block.rs (hunk 11 above)
Comment (rakita) [reply crates/rpc/rpc/src/eth/api/pending_block.rs:138]: Added

**Comment by rakita** [comment]
About `EnvWithHandlerCfg` and `CfgEnvWithHandlerCfg` I want to break those and remove `Env` soonish (After EOF) with generic tx/block https://github.com/bluealloy/revm/issues/916

I don't especially like these names, but want to have separated `HandlerCfg` from `Env` and it came up to this. I would look at them as temporary helpers for transitioning that are going to be modified/removed in few months. 

HandlerCfg allowed me to wrap Optimism changes and to have the same behaviour as previously.

**Review by mattsse — APPROVED** [review]
lgtm

ack on EOF

**Comment by Rjected** [comment]
needs a re review / approve from @emhane i think to merge

**Pull request merged**